File: CNSRC/cn-nt-cpp/exec.c
#define WINVER 0x0400
#define STRICT

#include <windows.h>
#include <string.h>

#define BUFSIZE         512U
#define eoErrIgnore     1
#define eoErrZero       2
#define eoAsync         4

DWORD Exec(LPCTSTR cmd, LPCTSTR redir, DWORD flags, DWORD opt)
{
 STARTUPINFO StartupInfo;
 PROCESS_INFORMATION ProcessInfo;
 SECURITY_ATTRIBUTES pipe_sa;
 HANDLE out_read = INVALID_HANDLE_VALUE;
 HANDLE out_write = INVALID_HANDLE_VALUE;
 HANDLE err_read = INVALID_HANDLE_VALUE;
 HANDLE err_write = INVALID_HANDLE_VALUE;
 HANDLE hfredir;
 DWORD Result;
 char Buf[BUFSIZE];
 DWORD BytesRead;
 DWORD BytesWrites;
 DWORD cmd_size;
 DWORD redir_size;

 cmd_size = strlen(cmd);
 OemToCharBuff((LPCSTR)cmd, (LPTSTR)cmd, cmd_size);

 redir_size = strlen(redir);
 OemToCharBuff((LPCSTR)redir, (LPTSTR)redir, redir_size);

 ZeroMemory(&StartupInfo, sizeof(STARTUPINFO));
 StartupInfo.cb = sizeof(StartupInfo);
 StartupInfo.dwFlags = STARTF_USESHOWWINDOW;
 StartupInfo.wShowWindow =  SW_SHOWNORMAL;
 if (((redir != NULL) && (*redir != 0)) &&
     ((opt & eoAsync) == 0))
  {
   /* Open pipes here */

   pipe_sa.nLength = sizeof(pipe_sa);
   pipe_sa.bInheritHandle = TRUE;
   pipe_sa.lpSecurityDescriptor = NULL;
   Buf[0] = 13; Buf[1] = 10;
   if (!CreatePipe(&out_read, &out_write, &pipe_sa, 0x8000))
    {
      out_read = INVALID_HANDLE_VALUE;
      out_write = INVALID_HANDLE_VALUE;
    }
   else WriteFile(out_write, Buf, 2, &BytesWrites, NULL);

   if (!CreatePipe(&err_read, &err_write, &pipe_sa, 0x8000))
    {
      err_read = INVALID_HANDLE_VALUE;
      err_write = INVALID_HANDLE_VALUE;
    }
   else WriteFile(err_write, Buf, 2, &BytesWrites, NULL);

   StartupInfo.wShowWindow = SW_HIDE;
   StartupInfo.dwFlags |= STARTF_USESTDHANDLES;
  }

 StartupInfo.hStdInput = INVALID_HANDLE_VALUE;
 StartupInfo.hStdOutput = out_write;
 StartupInfo.hStdError = err_write;
 flags |= NORMAL_PRIORITY_CLASS;
 if (!CreateProcess(NULL,         // pointer to name of executable module
                    cmd,          // pointer to command line string
                    NULL,         // pointer to process security attributes
                    NULL,         // pointer to thread security attributes
                    TRUE,         // handle inheritance flag
                    flags,        // creation flags
                    NULL,         // pointer to new environment block
                    NULL,         // pointer to current directory name
                    &StartupInfo, // pointer to STARTUPINFO
                    &ProcessInfo  // pointer to PROCESS_INFORMATION
                   )
    ) Result = 0;
 else Result = GetLastError();

 if (((Result == 0)||((opt & eoErrIgnore) != 0)) &&
     ((opt & eoAsync) == 0))
  {
   WaitForSingleObject(ProcessInfo.hProcess, INFINITE);
   CloseHandle(ProcessInfo.hThread );
   CloseHandle(ProcessInfo.hProcess );
   /* Create or open report file */

   hfredir = CreateFile(redir,                     // file to open
                       GENERIC_READ+GENERIC_WRITE, // open for reading
                       0,                          // do not share
                       NULL,                       // default security
                       OPEN_ALWAYS,                // existing or new file
                       FILE_ATTRIBUTE_NORMAL,      // normal file
                       NULL);                      // no attr. template

   if (hfredir != INVALID_HANDLE_VALUE)
    {
      /* Go to the end of file */
      SetFilePointer(hfredir, GetFileSize (hfredir, NULL), NULL, FILE_BEGIN);
      /* Copy output pipe to the report file */
      do
       {
         ReadFile(out_read, Buf, BUFSIZE, &BytesRead, NULL);
         WriteFile(hfredir, Buf, BytesRead, &BytesWrites, NULL);
       }
      while (BytesRead == BUFSIZE);
      /* Copy error pipe to the report file */
      do
       {
         ReadFile(err_read, Buf, BUFSIZE, &BytesRead, NULL);
         WriteFile(hfredir, Buf, BytesRead, &BytesWrites, NULL);
       }
      while (BytesRead == BUFSIZE);
      /* Close report file */
      CloseHandle(hfredir);
    };
  };
 if (err_read != INVALID_HANDLE_VALUE) CloseHandle(err_read);
 if (err_write != INVALID_HANDLE_VALUE) CloseHandle(err_write);
 if (out_read != INVALID_HANDLE_VALUE) CloseHandle(out_read);
 if (out_write != INVALID_HANDLE_VALUE) CloseHandle(out_write);
 if ((opt & eoErrZero) == 0) return Result;
 else return 0;
};

File: CNSRC/Sources/exec.pas
function ExecuteFile(FileName,StdInput: string;
                     TimeOut: integer;
                     var StdOutput:string) : boolean;

label Error;

type
  TPipeHandles = (IN_WRITE,  IN_READ,
                  OUT_WRITE, OUT_READ,
                  ERR_WRITE, ERR_READ);

type
  TPipeArray = array [TPipeHandles] of THandle;

var
  i         : integer;
  ph        : TPipeHandles;
  sa        : TSecurityAttributes;
  Pipes     : TPipeArray;
  StartInf  : TStartupInfo;
  ProcInf   : TProcessInformation;
  Buf       : array[0..1024] of byte;
  TimeStart : TDateTime;


function ReadOutput : string;
var
  i : integer;
  s : string;
  BytesRead : longint;

begin
  Result := '';
  repeat

    Buf[0]:=26;
    WriteFile(Pipes[OUT_WRITE],Buf,1,BytesRead,nil);
    if ReadFile(Pipes[OUT_READ],Buf,1024,BytesRead,nil) then
    begin
      if BytesRead>0 then
      begin
        buf[BytesRead]:=0;
        s := StrPas(@Buf[0]);
        i := Pos(#26,s);
        if i>0 then s := copy(s,1,i-1);
        Result := Result + s;
      end;
    end;

    if BytesRead <> 1024 then break;
  until false;
end;


begin
  Result := false;
  for ph := Low(TPipeHandles) to High(TPipeHandles) do
    Pipes[ph] := INVALID_HANDLE_VALUE;


  // Создаем пайпы
  sa.nLength := sizeof(sa);
  sa.bInheritHandle := TRUE;
  sa.lpSecurityDescriptor := nil;


  if not CreatePipe(Pipes[IN_READ],Pipes[IN_WRITE], @sa, 0 ) then
    goto Error;
  if not CreatePipe(Pipes[OUT_READ],Pipes[OUT_WRITE], @sa, 0 ) then
    goto Error;
  if not CreatePipe(Pipes[ERR_READ],Pipes[ERR_WRITE], @sa, 0 ) then
    goto Error;



  // Пишем StdIn
  StrPCopy(@Buf[0],stdInput+^Z);
  WriteFile(Pipes[IN_WRITE],Buf,Length(stdInput),i,nil);


  // Хендл записи в StdIn надо закрыть - иначе выполняемая программа
  // может не прочитать или прочитать не весь StdIn.

  CloseHandle(Pipes[IN_WRITE]);

  Pipes[IN_WRITE] := INVALID_HANDLE_VALUE;


  FillChar(StartInf,sizeof(TStartupInfo),0);
  StartInf.cb := sizeof(TStartupInfo);
  StartInf.dwFlags := STARTF_USESHOWWINDOW or STARTF_USESTDHANDLES;


  StartInf.wShowWindow := SW_SHOW; // SW_HIDE если надо запустить невидимо


  StartInf.hStdInput := Pipes[IN_READ];
  StartInf.hStdOutput := Pipes[OUT_WRITE];
  StartInf.hStdError := Pipes[ERR_WRITE];


  if not CreateProcess(nil, PChar(FileName), nil,
                       nil, True, NORMAL_PRIORITY_CLASS,
                       nil, nil, StartInf, ProcInf) then goto Error;

  TimeStart := Now;

  repeat
    Application.ProcessMessages;
    i := WaitForSingleObject(ProcInf.hProcess,100);
    if i = WAIT_OBJECT_0 then break;
    if (Now-TimeStart)*SecsPerDay>TimeOut then break;
  until false;


  if iWAIT_OBJECT_0 then goto Error;
  StdOutput := ReadOutput;

  for ph := Low(TPipeHandles) to High(TPipeHandles) do
    if Pipes[ph] <> INVALID_HANDLE_VALUE then
      CloseHandle(Pipes[ph]);


  CloseHandle(ProcInf.hProcess);
  CloseHandle(ProcInf.hThread);
  Result := true;
  Exit;


Error:

  if ProcInf.hProcessINVALID_HANDLE_VALUE then

  begin
    CloseHandle(ProcInf.hThread);
    i := WaitForSingleObject(ProcInf.hProcess, 1000);
    CloseHandle(ProcInf.hProcess);
    if iWAIT_OBJECT_0 then

    begin
      ProcInf.hProcess := OpenProcess(PROCESS_TERMINATE,
                                      FALSE,
                                      ProcInf.dwProcessId);

      if ProcInf.hProcess  0 then
      begin
        TerminateProcess(ProcInf.hProcess, 0);
        CloseHandle(ProcInf.hProcess);
      end;
    end;
  end;

  for ph := Low(TPipeHandles) to High(TPipeHandles) do
    if Pipes[ph]INVALID_HANDLE_VALUE then
      CloseHandle(Pipes[ph]);

end;